### ISSUE TITLE
[DeepLab v3+] How to set epochs?

### ISSUE THREAD

**Issue by SevenFoldDeep** [tensorflow/models#7340]
### Describe the problem
I have the model training with not problems so far. The thing I was wondering, is how do I determine the number of epochs? I only see the option to set the steps of the training as well as a weight decay after n steps. Is the training just carried out in steps rather than being divided into epochs of a certain number of steps each?

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
What is the top-level directory of the model you are using
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by SevenFoldDeep** [comment]
I have gone to Stackoverflow with this question.

**Comment by HaFred** [comment]
Hi. I have same puzzles, would you like to share your ideas?

**Comment by SevenFoldDeep** [comment]
> Hi. I have same puzzles, would you like to share your ideas?

I came to realize that epochs are simply the number of steps it takes for one complete pass through your training images, so you don't really need to set the epochs. 

Using DeepLab we can define the number of steps and the number of images per batch. So if I have 50 images, and a batch of 2 images per step, then it takes me 25 steps to go through all 50 images once. This is then one epoch. So if I want to train for 100 epochs I just need to set the number of steps in my training to 25*100=2500.

**Comment by HaFred** [comment]
> > Hi. I have same puzzles, would you like to share your ideas?
> 
> @HaFred
> I came to realize that epochs are simply the number of steps it takes for one complete pass through your training images, so you don't really need to set the epochs.
> 
> Using DeepLab we can define the number of steps and the number of images per batch. So if I have 50 images, and a batch of 2 images per step, then it takes me 25 steps to go through all 50 images once. This is then one epoch. So if I want to train for 100 epochs I just need to set the number of steps in my training to 25*100=2500.

Thank you for your reply. However, those pre-trained models under the repo, tensorflow/models/research, is already fixed right? 
As what you have mentioned, say we want to make it 50 epochs, can we assign num_steps=2500, batch_size=2 in the pipeline_config files to do so?

**Comment by SevenFoldDeep** [comment]
@HaFred
Yes the pre-trained models have some aspects which are fixed, but there are also many variables which you still have control over when fine-tuning (check the flags in the first part of train.py). 

You can manually set the number of steps by changing the integer in the 'training_number_of_steps' flag. Example:

flags.DEFINE_integer('training_number_of_steps', 30000,
                     'The number of steps used for training')

Here I set the number of training steps to 30000. For a batch size of 2 and a set of 50 images this would mean 1200 epochs.

I would recommend this blog post to see how you could set-up your training files, its the one I followed and it has a lot of great tips that helped me get DeepLab working on my dataset.

http://hellodfan.com/2018/07/06/DeepLabv3-with-own-dataset/

**Comment by HaFred** [comment]
> @HaFred
> Yes the pre-trained models have some aspects which are fixed, but there are also many variables which you still have control over when fine-tuning (check the flags in the first part of train.py).
> 
> You can manually set the number of steps by changing the integer in the 'training_number_of_steps' flag. Example:
> 
> flags.DEFINE_integer('training_number_of_steps', 30000,
> 'The number of steps used for training')
> 
> Here I set the number of training steps to 30000. For a batch size of 2 and a set of 50 images this would mean 1200 epochs.
> 
> I would recommend this blog post to see how you could set-up your training files as well as many other steps you need to follow when using deeplab on a new dataset. Its the one I followed and it has a lot of great tips that helped me get DeepLab working.
> 
> http://hellodfan.com/2018/07/06/DeepLabv3-with-own-dataset/

Thank you for your reply. It helps a lot. Indeed, some certain trees of tensorflow/models would be easier to implement, I am looking into that deeplab one, it is great.

**Comment by Raulpathak** [comment]
> @HaFred Yes the pre-trained models have some aspects which are fixed, but there are also many variables which you still have control over when fine-tuning (check the flags in the first part of train.py).
> 
> You can manually set the number of steps by changing the integer in the 'training_number_of_steps' flag. Example:
> 
> flags.DEFINE_integer('training_number_of_steps', 30000, 'The number of steps used for training')
> 
> Here I set the number of training steps to 30000. For a batch size of 2 and a set of 50 images this would mean 1200 epochs.
> 
> I would recommend this blog post to see how you could set-up your training files as well as many other steps you need to follow when using deeplab on a new dataset. Its the one I followed and it has a lot of great tips that helped me get DeepLab working.
> 
> http://hellodfan.com/2018/07/06/DeepLabv3-with-own-dataset/

Hello @SevenFoldDeep

**Comment by Raulpathak** [comment]
@SevenFoldDeep  i am  working on deeplab v3+  for lane detection , my detections are poor , I have trained with 36k images . i have not done any parameter tuning and training the model with 50k epocs.
Can u suggest some ways tom improve the detections , images I am gIving in the network are high resolution images

**Comment by SevenFoldDeep** [comment]
I would recommend looking at which layers you are keeping frozen from the pre-training. I would only unfreeze the final layers of the network. Other things you can try out is bigger batch size, different learning rate, learning rate scheduler. Also, are you training for 50K epochs or steps? It could be that you are training too long and the model is overfitting to your training data. Try training it for half the time and see how the results look. 

Also, since you have 36K images, try to train the model from scratch (no pre-training). Your dataset should be large enough for this.
